# Session #1 Homework

In [5]:
pip install wget

  Created wheel for wget: filename=wget-3.2-py3-none-any.whl size=9675 sha256=71611960be62f437aeb26fdecaac54607fd62c1f2910842cbd35611545043876
  Stored in directory: c:\users\darkcorp\appdata\local\pip\cache\wheels\04\5f\3e\46cc37c5d698415694d83f607f833f83f0149e49b3af9d0f38
Successfully built wget
Note: you may need to restart the kernel to use updated packages.


In [3]:
import numpy as np
import pandas as pd
import wget

### Question 1
What's the version of NumPy that you installed?

In [4]:
# Q1
np.__version__

'1.21.5'

### Getting the data
For this homework, we'll use the Car price dataset. Download it from here.

You can do it with wget:

`wget https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv`

Or just open it with your browser and click "Save as...".

Now read it with Pandas.

In [7]:
df = pd.read_csv('https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv')

In [8]:
df.head()

Make       Model  Year             Engine Fuel Type  Engine HP  \
0  BMW  1 Series M  2011  premium unleaded (required)      335.0   
1  BMW    1 Series  2011  premium unleaded (required)      300.0   
2  BMW    1 Series  2011  premium unleaded (required)      300.0   
3  BMW    1 Series  2011  premium unleaded (required)      230.0   
4  BMW    1 Series  2011  premium unleaded (required)      230.0   

   Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0               6.0            MANUAL  rear wheel drive              2.0   
1               6.0            MANUAL  rear wheel drive              2.0   
2               6.0            MANUAL  rear wheel drive              2.0   
3               6.0            MANUAL  rear wheel drive              2.0   
4               6.0            MANUAL  rear wheel drive              2.0   

                         Market Category Vehicle Size Vehicle Style  \
0  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
1                     Luxury,Performance      Compact   Convertible   
2                Luxury,High-Performance      Compact         Coupe   
3                     Luxury,Performance      Compact         Coupe   
4                                 Luxury      Compact   Convertible   

   highway MPG  city mpg  Popularity   MSRP  
0           26        19        3916  46135  
1           28        19        3916  40650  
2           28        20        3916  36350  
3           28        18        3916  29450  
4           28        18        3916  34500

In [12]:
df['Make']

0            BMW
1            BMW
2            BMW
3            BMW
4            BMW
          ...   
11909      Acura
11910      Acura
11911      Acura
11912      Acura
11913    Lincoln
Name: Make, Length: 11914, dtype: object

### Question 2
How many records are in the dataset?

Here you need to specify the number of rows.

* 16
* 6572
* **11914**
* 18990

In [19]:
df.describe().round(2)

Year  Engine HP  Engine Cylinders  Number of Doors  highway MPG  \
count  11914.00   11845.00          11884.00         11908.00     11914.00   
mean    2010.38     249.39              5.63             3.44        26.64   
std        7.58     109.19              1.78             0.88         8.86   
min     1990.00      55.00              0.00             2.00        12.00   
25%     2007.00     170.00              4.00             2.00        22.00   
50%     2015.00     227.00              6.00             4.00        26.00   
75%     2016.00     300.00              6.00             4.00        30.00   
max     2017.00    1001.00             16.00             4.00       354.00   

       city mpg  Popularity        MSRP  
count  11914.00    11914.00    11914.00  
mean      19.73     1554.91    40594.74  
std        8.99     1441.86    60109.10  
min        7.00        2.00     2000.00  
25%       16.00      549.00    21000.00  
50%       18.00     1385.00    29995.00  
75%       22.00     2009.00    42231.25  
max      137.00     5657.00  2065902.00

In [16]:
df.index

RangeIndex(start=0, stop=11914, step=1)

### Question 3
Who are the most frequent car manufacturers (top-3) according to the dataset?

* Chevrolet, Volkswagen, Toyota
* Chevrolet, Ford, Toyota
* Ford, Volkswagen, Toyota
* **Chevrolet, Ford, Volkswagen**

Note: You should rely on "Make" column in this question.

In [18]:
df['Make'].describe()

count         11914
unique           48
top       Chevrolet
freq           1123
Name: Make, dtype: object

In [41]:
df.groupby('Make').count().sort_values('Model', ascending=False).head(n=3)

Model  Year  Engine Fuel Type  Engine HP  Engine Cylinders  \
Make                                                                     
Chevrolet    1123  1123              1123       1117              1121   
Ford          881   881               881        868               881   
Volkswagen    809   809               809        809               805   

            Transmission Type  Driven_Wheels  Number of Doors  \
Make                                                            
Chevrolet                1123           1123             1123   
Ford                      881            881              881   
Volkswagen                809            809              809   

            Market Category  Vehicle Size  Vehicle Style  highway MPG  \
Make                                                                    
Chevrolet               622          1123           1123         1123   
Ford                    499           881            881          881   
Volkswagen              585           809            809          809   

            city mpg  Popularity  MSRP  
Make                                    
Chevrolet       1123        1123  1123  
Ford             881         881   881  
Volkswagen       809         809   809

### Question 4
What's the number of unique Audi car models in the dataset?

* 3
* 16
* 26
* **34**

In [26]:
df['Model'][
            (df['Make'] == 'Audi')
            ].nunique()

34

### Question 5
How many columns in the dataset have missing values?

* **5**
* 6
* 7
* 8

In [27]:
df.isnull().sum()

Make                    0
Model                   0
Year                    0
Engine Fuel Type        3
Engine HP              69
Engine Cylinders       30
Transmission Type       0
Driven_Wheels           0
Number of Doors         6
Market Category      3742
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

### Question 6
Find the median value of "Engine Cylinders" column in the dataset.

Next, calculate the most frequent value of the same "Engine Cylinders".

Use the fillna method to fill the missing values in "Engine Cylinders" with the most frequent value from the previous step.

Now, calculate the median value of "Engine Cylinders" once again.

Has it changed?

Hint: refer to existing mode and median functions to complete the task.

* Yes
* **No**

In [31]:
eng_cyl_median = df['Engine Cylinders'].median()
eng_cyl_median

6.0

In [32]:
eng_cyl_mode = df['Engine Cylinders'].mode()
eng_cyl_mode

0    4.0
Name: Engine Cylinders, dtype: float64

In [44]:
df['Engine Cylinders'].fillna(eng_cyl_mode).median()

6.0

### Question 7
1. Select all the "Lotus" cars from the dataset.
2. Select only columns "Engine HP", "Engine Cylinders".
3. Now drop all duplicated rows using drop_duplicates method (you should get a dataframe with 9 rows).
4. Get the underlying NumPy array. Let's call it X.
5. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
6. Invert XTX.
7. Create an array y with values [1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800].
8. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
9. What's the value of the first element of w?

Note: You just implemented linear regression. We'll talk about it in the next lesson.

* -0.0723
* **4.5949**
* 31.6537
* 63.5643

In [62]:
# Select all the "Lotus" cars from the dataset.
df_lotus = df[
                df['Make'] == 'Lotus'
]

# Select only columns "Engine HP", "Engine Cylinders".
# Now drop all duplicated rows using drop_duplicates method (you should get a dataframe with 9 rows).
# Get the underlying NumPy array. Let's call it X.
X = df_lotus[['Engine HP', 'Engine Cylinders']].drop_duplicates().values
X

array([[189.,   4.],
       [218.,   4.],
       [217.,   4.],
       [350.,   8.],
       [400.,   6.],
       [276.,   6.],
       [345.,   6.],
       [257.,   4.],
       [240.,   4.]])

In [64]:
# Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T.
XT = X.T
XT

array([[189., 218., 217., 350., 400., 276., 345., 257., 240.],
       [  4.,   4.,   4.,   8.,   6.,   6.,   6.,   4.,   4.]])

In [65]:
# Let's call the result XTX. 
XTX = XT.dot(X)
XTX

array([[7.31684e+05, 1.34100e+04],
       [1.34100e+04, 2.52000e+02]])

In [71]:
# Invert XTX.
XTX_inv = np.linalg.inv(XTX)
XTX_inv

array([[ 5.53084235e-05, -2.94319825e-03],
       [-2.94319825e-03,  1.60588447e-01]])

In [69]:
# Create an array y with values [1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800].
y = np.array([1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800])
y

array([1100,  800,  750,  850, 1300, 1000, 1000, 1300,  800])

In [74]:
# Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
w = XTX_inv.dot(XT).dot(y)
w

array([  4.59494481, -63.56432501])